features/101-design-webhooks/design.md

```diff
@@ -0,0 +1,188 @@
+
+```bash=
+curl -X POST -H "Content-Type: application/json" \
+-d @- http://localhost:3000/<gear name>/hooks <<EOF
+[
+    {
+        "events": ["<name of the event>", "<name of the event>"],
+        "url": "<url of the endpoint>",
+        "async": true|false,
+        "secret": "secret_string"
+    },
+    {
+        "events": ["<name of the event>", "<name of the event>"],
+        "url": "<url of the endpoint>",
+        "async": true|false,
+        "secret": "secret_string"
+    },
+]
+EOF
+[
+    {
+        "id": "<hook id>",
+        "events": ["<name of the event>", "<name of the event>"],
+        "url": "<url of the endpoint>",
+        "async": true|false,
+        "updated_at": "2019-01-06T20:39:23Z",
+        "created_at": "2019-01-06T17:26:27Z"
+    },
+    {
+        "id": "<hook id>",
+        "events": ["<name of the event>", "<name of the event>"],
+        "url": "<url of the endpoint>",
+        "async": true|false,
+        "updated_at": "2019-01-06T20:39:23Z",
+        "created_at": "2019-01-06T17:26:27Z"
+    },
+]
+```
+
+#### get hooks
+
+```bash=
+curl http://localhost:3000/<gear name>/hooks
+EOF
+[
+    {
+        "id": "<hook id>",
+        "events": ["<name of the event>", "<name of the event>"],
+        "url": "<url of the endpoint>",
+        "async": true|false,
+        "updated_at": "2019-01-06T20:39:23Z",
+        "created_at": "2019-01-06T17:26:27Z"
+    },
+    {
+        "id": "<hook id>",
+        "events": ["<name of the event>", "<name of the event>"],
+        "url": "<url of the endpoint>",
+        "async": true|false,
+        "updated_at": "2019-01-06T20:39:23Z",
+        "created_at": "2019-01-06T17:26:27Z"
+    },
+]
+```
+
+#### update a hook
+
+```bash=
+curl -X PATCH -H "Content-Type: application/json" \
+-d @- http://localhost:3000/<gear name>/hooks/<hook id> <<EOF
+{
+    "events": ["<name of the event>", "<name of the event>"],
+    "url": "<url of the endpoint>",
+    "async": true|false,
+    "secret": "secret_string"
+}
+EOF
+{
+    "id": "<hook id>",
+    "events": ["<name of the event>", "<name of the event>"],
+    "url": "<url of the endpoint>",
+    "async": true|false,
+    "updated_at": "2019-01-06T20:39:23Z",
+    "created_at": "2019-01-06T17:26:27Z"
+}
+```
+
+#### delete a hook
+
+```bash=
+curl -X DELETE -H "Content-Type: application/json" \
+http://localhost:3000/<gear name>/hooks/<hook id>/delete 
+EOF
+{
+     "result": "OK"
+}
+```
+
+## Verify Request
+
+When `secret` of a webhook is not empty, `X-Skygear-Webhhok-Signature` will be added to request headers, it is the HMAC hex digest of the POST request payload. The digest is generated using the sha256 hash function and the hook `secret` as the key.
+
+## Timeout of a `SYNC` hook
+
+A `SYNC` hook is considered failed if it can't response within 5 seconds.
+
+## Retry mechanism
+
+If a request got
+
+- 503 Services Unavailable
+- 429 Too Many Requests
+ 
+webhook will retry after 5 seconds, at most total 3 times.
```

Comment:
Is the retry for async only? Or both sync and async?

Reply:
I think for both sync and async should support retry.

Reply:
I would prefer for sync does not support retry.... that seems a bit contradict with the definition of "Sync" and the latency requirement